### PR TITLE
Noise

### PR DESCRIPTION
**Pull request by benma**

### CODE REVIEW

**Comment by x1ddos** [comment]
image

:disappointed:

**Comment by benma** [comment]
@x1ddos 

I added `bitbox02-noise` that is a separate package to provide the noise-protocol as used by the bitbox02 (noise is a framework, and we use a specific configuration of it). This crate also picks the crypto implementations. 

Best place to start looking at this is to compare noise.c processor:

https://github.com/digitalbitbox/bitbox02-firmware/blob/master/src/usb/noise.c#L248

with noise.rs processor:

https://github.com/digitalbitbox/bitbox02-firmware/pull/463/files#diff-f9ac5b760c4243dfabbe6949597f946dR69

The latter makes use of the `bitbox02-noise` crate to process the handshake and message processing (decrypt -> commander() -> encrypt).

New: the pairing is async, so you can cancel during the pairing code confirmation and the device will cancel it.

Also new: -50kb binary size, probably a bit more to come when deleting the rest of noise-c.

**Comment by benma** [comment]
> image
> 
> disappointed

image

Click on the commits :joy:

**Comment by benma** [comment]
Missing in this PR: [this bitbox base pairing shortcut](https://github.com/digitalbitbox/bitbox02-firmware/blob/master/src/usb/noise.c#L299). I'll save it for another PR, as it's not quite clear how to handle it.

File: src/rust/bitbox02-noise/src/handshake.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module implements the X25519 trait needed by noise_protocol
+//! by using the x25519_dalek crate. It is adapted from
+//! https://github.com/sopium/noise-rust/blob/2e4bad0dbe3c674d4e1820e2325ad108c1150dfb/noise-rust-crypto/src/lib.rs#L29,
```
Comment (benma) [inline src/rust/bitbox02-noise/src/handshake.rs:17]: need to update this to their new version: https://github.com/sopium/noise-rust/blob/76fb694f06b429879c264087f496958a99710356/noise-rust-crypto/src/lib.rs#L31

File: src/rust/bitbox02-noise/src/noise_xx.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+    enum MockRandom32 {}
+    impl Random32 for MockRandom32 {
+        fn mcu_32_bytes(out: &mut [u8; 32]) {
+            out.copy_from_slice(b"llllllllllllllllllllllllllllllll")
+        }
+    }
+
+    #[test]
+    pub fn test_full() {
```
Comment (benma) [inline src/rust/bitbox02-noise/src/noise_xx.rs:251]: will add more complete test coverage (negative tests) if it looks like the code won't change much anymore.

**Comment by x1ddos** [comment]
I wonder if it's somehow possible to skip importing benches and tests of various vendored crates. We don't use them and it's really a lot of files, including test data blobs.

I don't think we use `rand_core` crate? Looks like it's an optional dep of chacha20. Possible to disable that?

Also, would've updated `binascii` in a separate commit.

**Comment by benma** [comment]
> I don't think we use `rand_core` crate? Looks like it's an optional dep of chacha20. Possible to disable that?

rand-core is a non-optional dependency of `x25519-dalek`, which uses the traits therein to get the randomness source for key generation.

As you see in `bitbox02-noise/src/x25519.rs`, the `genkey()` is actually not calling `x25519-dalek`, so technically we do not use it. We could also implement the rand-core traits and then call the `x25519-dalek` genkey.

**Comment by x1ddos** [comment]
Just having looked at few files I can already say this Rust version is somehow much easier to read and follow.

File: src/rust/bitbox02-rust/Cargo.toml [hunk 3]
```diff
@@ -27,7 +27,8 @@ doctest = false
 
 [dependencies]
 bitbox02 = {path = "../bitbox02"}
-binascii = { version = "0.1.4", default-features = false, features = ["encode"] }
+bitbox02-noise = {path = "../bitbox02-noise"}
+binascii = "0.1"
```
Comment (x1ddos) [inline src/rust/bitbox02-rust/Cargo.toml:31]: Hm... why did this change?

File: src/rust/bitbox02-rust-c/src/noise.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// `private_key_out` must be 32 bytes.
+#[no_mangle]
+pub extern "C" fn rust_noise_generate_static_private_key(
+    mut private_key_out: crate::util::BytesMut,
+) {
+    let key = bitbox02_noise::generate_static_private_key::<bitbox02_rust::hww::BB02Random32>();
```
Comment (x1ddos) [inline src/rust/bitbox02-rust-c/src/noise.rs:20]: This reads weird, as if there's an independent hardware random generator, imho.
Why not use directly `bitbox02_rust::hww::noise::BB02Random32`?

**Comment by x1ddos** [comment]
> > I don't think we use `rand_core` crate? Looks like it's an optional dep of chacha20. Possible to disable that?
> 
> rand-core is a non-optional dependency of `x25519-dalek`, which uses the traits therein to get the randomness source for key generation.
> 
> As you see in `bitbox02-noise/src/x25519.rs`, the `genkey()` is actually not calling `x25519-dalek`, so technically we do not use it. We could also implement the rand-core traits and then call the `x25519-dalek` genkey.

Ah, if it's non-optional then it stays as a dependency anyway. I wanted to remove it from the deps completely but ok. Maybe we'll patch x25519-dalek in the future to somehow make it optional.

File: src/rust/Cargo.lock [hunk 5]
```diff
@@ -1,166 +1,373 @@
...
- "bitbox02-sys 0.1.0",
- "util 0.1.0",
- "zeroize 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)",
+ "bitbox02-sys",
+ "util",
+ "zeroize",
+]
+
+[[package]]
+name = "bitbox02-noise"
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/Cargo.lock:36]: the deps only come with adding this. i moved all the dep stuff into one commit at the end of the process.

didn't feel like spending time making a separate commit with an empty / bootstrap bitbox02-noise to make it consistent.

File: src/rust/bitbox02-noise/src/x25519.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    fn name() -> &'static str {
+        "25519"
+    }
+
+    fn genkey() -> Self::Key {
+        let mut k = Self::Key::new();
+        R::mcu_32_bytes(k.deref_mut());
+        k[0] &= 248;
+        k[31] &= 127;
+        k[31] |= 64;
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/x25519.rs:47]: The top of the file has a comment that links to the file where the whole thing was taken from. Not enough?

File: src/rust/bitbox02-rust/src/hww/noise.rs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::vec::Vec;
+use bitbox02::memory;
+use core::cell::RefCell;
+
+const OP_I_CAN_HAS_HANDSHAEK: u8 = b'h';
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/noise.rs:19]: yeah had to fix it, handshake is not proper catspeak :D

File: src/rust/bitbox02-rust/src/hww/noise.rs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+                    usb_out.push(1); // let app know we do require verification
+                }
+                Ok(())
+            }
+        },
+        Some((&OP_I_CAN_HAS_PAIRIN_VERIFICASHUN, b"")) => {
+            let hash = state.get_handshake_hash()?;
+            if crate::workflow::pairing::confirm(&hash).await {
+                state.set_pairing_verified()?;
+                memory::add_noise_remote_static_pubkey(&state.remote_static_pubkey()?)?;
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/noise.rs:99]: i think either way is fine, i have not seen any instance of memory storing failing, and it if fails there might be much bigger problems. maybe even panic makes sense?

File: src/rust/bitbox02-rust/src/hww/mod.rs [hunk 9]
```diff
@@ -75,7 +76,9 @@ pub async fn process_packet(usb_in: Vec<u8>) -> Vec<u8> {
         return Vec::new();
     }
 
-    // -- Process anything not ported to Rust yet. --
-    // This function is blocking.
-    bitbox02::hww::process_packet(usb_in)
+    let mut out = [OP_STATUS_SUCCESS].to_vec();
+    match noise::noise(usb_in, &mut out).await {
+        Ok(()) => out,
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/mod.rs:84]: I had a version before where the vector would be returned and appended to OP_STATUS_SUCCESS, similar to your suggestion.

But that actually forces a copy of the whole response, which can be quite big.

This is why I made the function append to the vector instead.

File: src/rust/bitbox02-rust/Cargo.toml (hunk 3 above)
Comment (benma) [reply src/rust/bitbox02-rust/Cargo.toml:31]: wrong rebase i assume from when I bumped binascii in the other PR. will fix.

File: src/rust/bitbox02-noise/src/x25519.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02-noise/src/x25519.rs:47]: That link may become dead and their code doesn't explain it for us, mere mortals, either anyway. I'd prefer to be explicit - it doesn't cost much, does it.

File: src/rust/bitbox02-rust/src/hww/mod.rs [hunk 10]
```diff
@@ -75,7 +76,9 @@ pub async fn process_packet(usb_in: Vec<u8>) -> Vec<u8> {
...
+        Err(noise::Error) => [OP_STATUS_FAILURE].to_vec(),
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/mod.rs:85]: So both clients just check that the response is `OP_STATUS_SUCCESS`:

https://github.com/digitalbitbox/bitbox02-firmware/blob/6be39e5ff4288d984ece2e9895a2efbb18edd160/py/bitbox02/bitbox02/communication/bitbox_api_protocol.py#L291

https://github.com/digitalbitbox/bitbox02-api-go/blob/741357feee4ae0fc3bf7fcbd72c2f30168062448/api/firmware/query.go#L123

So the require_pairing_verification response was never used.

Do you think I should add it back anyway?

> which I believe translates to hwwRspNack.

It doesn't, `hwwRsp*` are on a different layer (it's the top level arbitration framing).

[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02-rust/src/hww/mod.rs:85]: Ah ok. I was looking in the wrong place.
Then I think it's fine as is. The pairing is enforced anyway. I just worried about compatibility with the go and py packages.
:+1:

File: src/rust/bitbox02-noise/src/noise_xx.rs [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    ///
+    /// See also: [noiseexplorer.com/patterns/XX](https://noiseexplorer.com/patterns/XX/).
+    ///
+    /// Returns: `Response(<respB>)` (see above) for the 1st handshake message, and `Done` for the
+    /// 2nd handshake message.
+    pub fn handshake(&mut self, msg: &[u8]) -> Result<HandshakeResult, Error> {
+        use core::convert::TryInto;
+        match self {
+            State::Initialized(handshake_state) => {
+                let payload = handshake_state.read_message_vec(msg)?;
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/noise_xx.rs:121]: It can also fail just due to an invalid message be the app? 

Invalid noise state can happen at anytime, e.g. if one of the sides misses a a round or sth like that.

The app can always reset the noise state by starting a new nosie channel by sending `OP_I_CAN_HAS_HANDSHAEK`.

**Comment by benma** [comment]
@x1ddos thanks for all comments. I pushed one commit with a link to your comment in the commit msg, so it's easier to match up the changes. Will squash before merge.

File: src/rust/bitbox02-rust-c/src/noise.rs (hunk 4 above)
Comment (benma) [reply src/rust/bitbox02-rust-c/src/noise.rs]: Done

File: src/rust/bitbox02-rust/src/hww/noise.rs [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        Some((&OP_HER_COMEZ_TEH_HANDSHAEK, rest)) => match state.handshake(rest)? {
+            bitbox02_noise::HandshakeResult::Response(msg) => {
+                usb_out.extend(msg);
+                Ok(())
+            }
+            bitbox02_noise::HandshakeResult::Done => {
+                let already_verified =
+                    memory::check_noise_remote_static_pubkey(&state.remote_static_pubkey()?);
+                if already_verified {
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/noise.rs:86]: Done

File: src/rust/bitbox02-rust/src/hww/noise.rs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+///   - Enforce pairing confirmation if the remote party is seen for the first time
+///   - Remote party can invoke pairing confirmation anytime after the handshake
+/// - Decrypt, process, encrypt
+///
+/// The result is appended to `usb_out`.
+///
+/// Returns Err if anything goes wrong:
+/// - Invalid OP-code
+/// - Noise message in the wrong state (e.g. handshake before init, etc.).
+pub async fn noise(usb_in: Vec<u8>, usb_out: &mut Vec<u8>) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/noise.rs]: Done

File: src/rust/bitbox02-noise/src/noise_xx.rs [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+                ..
+            } => Err(Error::PairingVerificationRequired),
+
+            State::Ready {
+                pairing_verification_required: false,
+                send,
+                ..
+            } => {
+                let start = out.len();
+                let encrypted_len = msg.len() + 16;
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/noise_xx.rs:213]: Done

File: src/rust/bitbox02-noise/src/noise_xx.rs [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    /// Can be called at any time to start waiting for a new communication channel.
+    ///
+    /// `static_private_key` is the local static key. It can be generated using
+    /// `generate_static_private_key()`.
+    pub fn init(&mut self, static_private_key: Sensitive<PrivateKey>) {
+        let hs = HandshakeState::new(
+            noise_protocol::patterns::noise_xx().clone(),
+            false,
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/noise_xx.rs]: Done

File: src/rust/bitbox02-noise/src/x25519.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/x25519.rs:52]: > Would be slightly easier to read if these were in binary like `0b11111000`

Won't change that as it is used like this also in x25519-dalek.

Added this comment now:

```
// Copied from: https://github.com/sopium/noise-rust/blob/76fb694f06b429879c264087f496958a99710356/noise-rust-crypto/src/lib.rs#L49-L51
        // which in turn copied it from:
        // https://github.com/dalek-cryptography/x25519-dalek/blob/ecd6be674850a99ad26404f6aa29b0cf79642b97/src/x25519.rs#L162-L164
        // which is also in our vendored deps: `vendor/x25519-dalek/src/x25519.rs`.
```

even if the links become dead, the pointer to the vendored file will remain.

[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02-noise/src/x25519.rs:52]: When I copy code it becomes something I own and have to maintain regardless of the original, so I'd want to understand exactly what it does, esp. after 6 month or a year. "Won't change that as it is used like this also in x25519-dalek." wouldn't really justify it for me. Do you always keep code exactly as is when copied from stackoverflow? :)

I guess the pointer to the vendored version will do. Thanks!

File: src/rust/bitbox02-noise/src/noise_xx.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/noise_xx.rs:121]: On the other hand probably it won't hurt to reset just in case.

[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/noise_xx.rs]: Done

File: src/rust/bitbox02-rust/src/hww/noise.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02-rust/src/hww/noise.rs:100]: panic is a bit harsh imho. if a channel is established ok and we just can't persist to skip pairing confirmation, why should users suffer more than necessary, theoretically. I mean, if this is now required then sure. But it doesn't seem like persisting pairing confirmation is actually a prerequisite to safely communicate with the device, is it.   

> i have not seen any instance of memory storing failing

doesn't mean it won't at some point! :)

File: src/rust/bitbox02-noise/src/x25519.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-noise/src/x25519.rs:52]: > wouldn't really justify it for me. Do you always keep code exactly as is when copied from stackoverflow? :)

> I'd want to understand exactly what it does, esp. after 6 month or a year. 

If you change the code but the link to the source is still there, it is extra confusing and takes more time to re-evaluate it.

Besides, there is really no reason to change a cryptographic key generation function from a crypto source. There is no benefit, but a non zero risk of introducing an error.

File: src/rust/bitbox02-rust/src/hww/noise.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/noise.rs:100]: > I mean, if this is now required then sure. But it doesn't seem like persisting pairing confirmation is actually a prerequisite to safely communicate with the device, is it.

No, but not being able to persist data could mean all sorts of weird issues? That's why I don't think a panic is so harsh.

But in this case I'll change it to be the same as before.

[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/hww/noise.rs:100]: Done

[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02-rust/src/hww/noise.rs]: > No, but not being able to persist data could mean all sorts of weird issues? That's why I don't think a panic is so harsh.

By that definition the firmware should probably panic in many other places!

File: src/rust/bitbox02-noise/src/x25519.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02-noise/src/x25519.rs:52]: Oh sure, I'm fine with leaving it in decimal. I was mostly complaining about:

> The top of the file has a comment that links to the file where the whole thing was taken from. Not enough?

because imho it isn't enough: comments like "copied from here and there" with links which don't explain what it's for and why make it harder to understand and maintain in the future.

File: src/rust/bitbox02-noise/src/noise_xx.rs [hunk 16]
```diff
@@ -213,6 +213,7 @@ impl<R: Random32> State<R> {
                 let encrypted_len = msg.len() + 16;
                 // Make space for result.
                 out.resize(start + encrypted_len, 0);
+                // This also adds the MAC.
```
Comment (x1ddos) [inline src/rust/bitbox02-noise/src/noise_xx.rs:222]: please consider adding another comment on the line 213 above `... + 16; // extra 16 bytes for MAC`.

File: src/rust/bitbox02-rust/src/hww/noise.rs [hunk 17]
```diff
@@ -66,7 +66,7 @@ impl core::convert::From<()> for Error {
 /// Returns Err if anything goes wrong:
 /// - Invalid OP-code
 /// - Noise message in the wrong state (e.g. handshake before init, etc.).
-pub async fn process(usb_in: Vec<u8>, usb_out: &mut Vec<u8>) -> Result<(), Error> {
+pub(crate) async fn process(usb_in: Vec<u8>, usb_out: &mut Vec<u8>) -> Result<(), Error> {
```
Comment (x1ddos) [inline src/rust/bitbox02-rust/src/hww/noise.rs:69]: I don't think this is needed for `rust_noise_generate_static_private_key`. Why the change?

Comment (benma) [reply src/rust/bitbox02-rust/src/hww/noise.rs:69]: `pub(crate)` means it's only public within the crate, so hww.rs can access it, but not the `-c` crate.

Comment (x1ddos) [reply src/rust/bitbox02-rust/src/hww/noise.rs:69]: oh, of course. because it's using `bitbox02_rust::hww::noise::BB02Random32`. good point.

File: src/rust/bitbox02-noise/src/noise_xx.rs (hunk 16 above)
Comment (benma) [reply src/rust/bitbox02-noise/src/noise_xx.rs:222]: Done
